### DISCUSSION TITLE
Key generation for asymmetric encryption algorithm

### DISCUSSION THREAD

**Discussion by ranjithravitk** [provider-corner/.github#8] [Q&A]
I am implementing the Elgamal encryption scheme as a provider. While implementing, I came across the 
int OSSL_FUNC_asym_cipher_encrypt_init(void *ctx, void *provkey,const OSSL_PARAM params[]); here the provkey said to be consisting of a key object that needs to be generated using provider-keymgmt functions. I have two questions
i) Does openSSL automatically generate the key object before encryption if I implement the keymgmt functions?
ii) Does the key object need to in EVP_PKEY structure format? if so I couldn't find the EVP_PKEY structure for Elgamal, where can I find it or how can I proceed further? any input will be helpful. @levitte apologies for tagging you. But I don't know whether the discussion board is active.

**Comment by levitte** [comment]
This are really questions for the OpenSSL project itself, but I'm game giving some answers here.

1. OpenSSL will make sure that they key object that is used is made by the project-keymgmt of the same provider to which the OSSL_FUNC_asym_cipher_encrypt_init() call is made.  That way, it's guaranteed that the provider knows the key object structure.
2. The key object may have *any structure you want*, that's entirely up to you.  OpenSSL doesn't know, and doesn't care, it simply passes the *provkey* pointer around.
 
For your information, OpenSSL doesn't implement ElGamal, so for all intents and purposes, an ElGamal implementation adds something new and unknown to OpenSSL.  Cool!  I'll happily add a link to your project if you set one up.

**Reply by ranjithravitk** [reply]
Hi @levitte thanks for replying. It helped me to understand it. I am working on the implementation and message here in the same thread once done. Thank you so much!
